consensus/src/iteration_ctx.rs

```diff
@@ -0,0 +1,203 @@
+        let step = StepName::Proposal.to_step(iter);
+        self.get_committee(step)
+            .and_then(|c| c.iter().next().map(|p| *p.bytes()))
+    }
+
+    pub(crate) fn get_validation_committee(
+        &self,
+        iter: u8,
+    ) -> Option<&Committee> {
+        let step = StepName::Validation.to_step(iter);
+        self.get_committee(step)
+    }
+
+    pub(crate) fn insert(&mut self, step: u16, committee: Committee) {
+        self.committees.insert(step, committee);
+    }
+}
+
+/// Represents a shared state within a context of the execution of a single
+/// iteration.
+pub struct IterationCtx<DB: Database> {
+    validation_handler: Arc<Mutex<validation::handler::ValidationHandler>>,
+    ratification_handler:
+        Arc<Mutex<ratification::handler::RatificationHandler>>,
+    proposal_handler: Arc<Mutex<proposal::handler::ProposalHandler<DB>>>,
+
+    pub join_set: JoinSet<()>,
+
+    round: u64,
+    iter: u8,
+
+    /// Stores any committee already generated in the execution of any
+    /// iteration of current round
+    pub(crate) committees: RoundCommittees,
+
+    /// Implements the adaptive timeout algorithm
+    step_base_timeout: Duration,
+}
+
+impl<D: Database> IterationCtx<D> {
+    pub fn new(
+        round: u64,
+        iter: u8,
+        proposal_handler: Arc<Mutex<proposal::handler::ProposalHandler<D>>>,
+        validation_handler: Arc<Mutex<validation::handler::ValidationHandler>>,
+        ratification_handler: Arc<
+            Mutex<ratification::handler::RatificationHandler>,
+        >,
+        round_base_timeout: Duration,
+    ) -> Self {
+        Self {
+            round,
+            join_set: JoinSet::new(),
+            iter,
+            proposal_handler,
+            validation_handler,
+            ratification_handler,
+            committees: Default::default(),
+            step_base_timeout: round_base_timeout,
+        }
+    }
+
+    /// Executed on starting a new iteration, before Proposal step execution
+    pub(crate) fn on_begin(&mut self, iter: u8) {
+        self.iter = iter;
+    }
+
+    /// Executed on closing an iteration, after Ratification step execution
+    pub(crate) fn on_close(&mut self) {
+        debug!(
+            event = "iter completed",
+            len = self.join_set.len(),
+            round = self.round,
+            iter = self.iter,
+        );
+        self.join_set.abort_all();
+    }
+
+    /// Handles an event of a Phase timeout
+    pub(crate) fn on_timeout_event(&mut self) {
+        self.increase_step_timeout();
+    }
+
+    /// Calculates and returns the adjusted timeout for the specified step
+    pub(crate) fn get_timeout(&self, step_name: StepName) -> Duration {
+        match step_name {
+            StepName::Proposal => self.step_base_timeout,
+            StepName::Validation => self.step_base_timeout * 3,
+            StepName::Ratification => self.step_base_timeout * 3,
```

Comment:
Why this `*3`?

IIUC the iteration timout will increase to 35secs (5+15+15). A bit too much high IMO

Reply:
This is as per requirement:
```
the calculation should take into account the expected number of messages to be received. For instance, since the Generation phase only requires one message to be received, we can set GENERATION_TIMEOUT=STEP_BASE_TIMEOUT; instead, for the voting phases, which require several messages to be received, we can set REDUCTION1_TIMEOUT=3*STEP_BASE_TIMEOUT.
 ```

Reply:
> Why this `*3`?
> 
> IIUC the iteration timout will increase to 35secs (5+15+15). A bit too much high IMO

The `*3` was chosen arbitrarily as a matter of example. We should come up with a more sensible multiplier.
AFAIU this will be addressed with a follow-up issue.